Rules and groups (#211)
hiero-ledger/governance: README.md

Reviewer: Why is this in the governance `README.md` file to begin with? This should be framed as an amendment/addendum to the Technical Charter.

Reviewer 2: Should it be there? Should it be in the top level contributors file?

Reviewer 3: Yes, the role follows the person.

# Hiero technical governance

This repository contains the `config.yaml` file that is used to configure all GitHub repositories of the organisation.

> https://clowarden.io is used as tool to manage the GitHub ressources and the changelist for all Hiero repos can be found [here](https://clowarden.io/audit/?organization=LFDT-Hiero).

Author: Outdated

## Definition of general roles

In general the Hiero project (and the LFDT) defines 3 different types of **roles**: **Contributors**, **Committers**, and **Maintainers**.
Independent of the role each person who contributes to Hiero must follow our **Code of Conduct** (https://github.com/hiero-ledger/.github/blob/main/CODE_OF_CONDUCT.md).
Those roles are not assigned globally for the Hiero project but per **project**.
In most cases a project is a single GitHub repository, but it can also be a group of repositories that are related to each other.
A person can have multiple roles in multiple projects.

### Contributors
In general any person that does any contribution to the project is a Contributor.

> A contribution can be anything from opening an issue, to writing code, to writing documentation, etc.

Reviewer 4: I recommend the following:

A contribution can be anything from opening an issue, to writing code, to writing documentation, etc:
- If a PR has gone 2 days without approvals or an Issue or Discussion has no activity for 2 days, the owner must reach our to our hiero-general Discord channel.
- If a PR or discussion has gone stale for 2 weeks, the repo maintainers can close them due to inactivity.

We honor all contributions to the project, and we want to make sure that all contributions are recognized.
If a person is a Contributor and has made significant contributions to the project, they can be invited to become a Committer.
Any Committer or Maintainer of the given project can nominate a person to become a Committer.
All Maintainers should vote on that nomination.

> The vote should be held in a GitHub PR (by doing a change to the `config.yaml` file).

Author: Agree but we do not have discussed that in the open ;) Once this PR is merged In will open an discussion for it

> The vote should be held in a GitHub PR (by doing a change to the `config.yaml` file).

Reviewer: We should incorporate it in this open PR and seek approvals; otherwise, this is an incomplete document. I don't see any benefit to opening a discussion thread since a PR in and of itself is a public discussion.

> The vote should be held in a GitHub PR (by doing a change to the `config.yaml` file).

Reviewer 2: I agree with this. Voting should be defined as part of this initial document to provide complete and clear direction to teams when setting up maintainers and committers.

> The vote should be held in a GitHub PR (by doing a change to the `config.yaml` file).

Reviewer 4: I propose changing line 18 t 21 to:

Any Committer or Maintainer of the given project should nominate a Contributor to become a Committer based on the project's needs and following these steps:

- **Define your project's needs:** A repository should have minimum 3 committers to ensure a project's sustainability and resilience. It will also allow for fairness in the votes. Project owners, can choose to have more committers depending on the size of their project.
- **Evaluate your project's performance:** Committers and Maintainers should monitor their project health and velocity and make decisions about whether the project needs more Committers.
- **Verify your project's needs:** Committers or Maintainers should define the number of good contributions that a Contributor must reach to consider a nomination. This is based on the size and complexity of the repo in question. 
- **Nominate more resources:** To promote a contributor, a vote should be held in a GitHub PR (by doing a change to the `config.yaml` file) and by making mention of the basis reached for nominating the Contributor.
- **Approve the nomination:** All Maintainers should vote on that nomination and the vote passes when reaching majority of votes. 
- **Make adjustments:** Maintainers should be responsible for removing inactive committers. A committer is considered inactive if there has been no activity for six months and there has been no response when trying to contact them.

### Committers
A Committer has specific rights for the project.
As an example a Committer can assign people to issues, merge PRs, or create branches directly in a repository of the project.
A detailed overview of all rights will be given below.
As long as a Committer follows our Code of Conduct, makes reasonable contributions to the project,
and is not violating any of our policies, they will keep their Committer role.
If a Committer is not making any contributions to the project anymore for a longer periode of time they will lose their Committer role.

> We need to define a reasonable timeframe for that.

Author: Let's change it to "a reasonable timeframe for that must be defined" and discuss after PR is merged if it is by project (defined by maintainers) or global for Hiero. I'm fine with both but would suggest a best practice for case 1 where projects can rely on. Anyway having an open discussion will be helpful.

> We need to define a reasonable timeframe for that.

Reviewer 2: "a reasonable timeframe that must be defined." The for there is erroneous.

> We need to define a reasonable timeframe for that.

Reviewer 4: From the research that i been doing a committer should be demoted if:
- Breaks code of conduct
- Shown no activity for 6 months and does not responds when reached
- Keeps making bad decisions (their approvals and contributions are actively affecting the project's performance).

Each project should have 1-3 Maintainers.
A Maintainer should drive the project's roadmap/vision and take care of topics like community meetings,
the general structure of the project (fits our best practices, is secure, ...), and so on.
In addition, the Maintainers should report project updates to the Hiero **Technical Steering Committee** (TSC).
A Committer of a project can be nominated to become a Maintainer of that project, and all Maintainers should vote on that.
The vote should be held in a GitHub PR (by doing a change to the `config.yaml` file).
It is not the goal that every Committer of a project becomes a Maintainer.
Having hundreds of Maintainer for a single project does not make sense.
As already defined for the Committer role, a Maintainer can lose the Maintainer role if they harm the project in any way.
For a Maintainer the contribution to the project should be on a really high level.
Therefore, it is a valid decision to remove a Maintainer from the project and give that person the Committer role.
Remove a Maintainer from the project and give that person the Contributors role should only happen in

> really extreme cases where the Maintainer is harming the project or ignoring the Code of Conduct.

Author: Outdated

## Definition of groups

Since Hiero contains multiple projects, we need to define GitHub **groups** that are used to manage the access to the GitHub repositories.
The groups are defined in the `config.yaml` file and are used to manage the access to the GitHub repositories that belong to a project.
Contributors do not have any specific rights, and therefore we do not need to define any groups for them.

> Committers** and Maintainers are defined in groups that are used to manage the access to the GitHub repositories.

Reviewer 5: ```suggestion
Committers and Maintainers are defined in groups that are used to manage the access to the GitHub repositories.
```

For each project we have defined a Committers group and a Maintainers group: 

- **`PROJECT-maintainers`**: For each project we have a group that defines the Maintainers of that project.
- **`PROJECT-committers`**: For each project we have a group that defines the Committers of that project.

Next to that you can find some other general groups in the `config.yaml` file:

- **`hiero-automation`**: TODO

> - **`hiero-mirror-node-automation`**: TODO

Reviewer 2: ```suggestion
- **`hiero-mirror-node-automation`**: CI/CD Automation serice account for hiero-ledger mirror node projects
```

- **`github-maintainers`**: The group is responsible for maintaining our GitHub environment and infrastructure (including GitHub Action runners, ...).
- **`community`**: A group that contains all people that are currently not part of any other (non-temporary) group.

>   We use this temporarly to invite people to the Hiero org.

Reviewer 2: ```suggestion
  We use this temporarily to invite people to the Hiero org.
```

- **`github-committers`**:
- **`community`**: A group that contains all people that are currently not part of any other (non-temporary) group.

>   We use this temporarly to invite people to the Hiero org.

Reviewer 4: The community group is gonna be removed:
https://github.com/hiero-ledger/governance/pull/215

> - **`security-maintainers`**: TODO
> - **`prod-security`**: TODO
> - **`sec-ops`**: TODO

Reviewer 2: ```suggestion
- **`security-maintainers`**: This group contains members of the security teams who are responsible for monitoring and maintaining security within hiero-ledger projects
- **`prod-security`**: This group contains members of the security teams who are responsible for monitoring and maintaining security within hiero-ledger projects
- **`sec-ops`**: This group contains members of the security teams who are responsible for monitoring and maintaining security within hiero-ledger projects
```

## Definitions of rights

For each project the following GitHub roles should be defined for the groups that work on the project.
Those GitHub roles defines the rights of the groups on the GitHub repositories of the project.
All roles are based on the [GitHub role defintions](https://docs.github.com/en/organizations/managing-user-access-to-your-organizations-repositories/managing-repository-roles/repository-roles-for-an-organization).

- **`github-maintainers`**: `MAINTAIN`

> - **`tsc`**: `ADMIN`

Reviewer 2: ```suggestion
- **`tsc`**: `MAINTAIN`
```

- **`PROJECT-maintainers`**: `MAINTAIN`
- **`PROJECT-committers`**: `WRITE`

The GitHub roles are assigned in the `config.yaml`.

> ## Usage of GitHub CODEOWNERS in combination with our roles, groups,and rights

Author: ```suggestion
## Usage of GitHub CODEOWNERS in combination with our roles, groups, and rights
```

A project can decide to use GitHub `CODEOWNERS` (see https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners).
Using `CODEOWNERS` is a best practice.

> It should be encouraged in all projects (certainly not required).

Reviewer 5: ```suggestion
It should be encouraged in all projects but is not required.
```

This helps maintainers ensure that the codebase is getting reviewed by the experts on those areas of the code.
Combining the Contributors, Committers, and Maintainers roles with GitHub `CODEOWNERS` is a little bit tricky.
So far, the following definitions are valid:

- The `CODEOWNERS` file defines some specific groups that are **codeowner** of particular files (or folders) of the repository.
- A PR that contains changes on those files needs to be approved by one of those codeowners.
- A `CODEOWNERS` file will not affect the definition of Contributors, Committers, and Maintainers.

> - With that in mind, it makes zero sense that a group in the `CODEOWNERS` file is not a Committer, and Maintainer group
>   (see **`PROJECT-maintainers`** and **`PROJECT-committers`**) or one of our special groups like **`lf-staff`**.

Reviewer 2: ```suggestion
- The minimum permission a **codeowner** must have is `write` on a given repository. As such a **codeowner** must be a Committer or Maintainer on the project
  (see **`PROJECT-maintainers`** and **`PROJECT-committers`**) or one of our special groups like **`lf-staff`**.
```
